main.cpp

```diff
@@ -41,22 +41,458 @@ If you need inspiration for what to write, take a look at previously approved st
+    int cupsOfTeaAvailable;
+    int participants = 3;
+    bool tooMuchTea = false;
+    bool biscuits = false;
+    bool rain = false;
+ };
+
+int TeaParty::serve(int cups)
+{
+    if(!tooMuchTea || (cupsOfTeaAvailable < 3))
+    {
+    cout << "serve " << cups << " cups of tea"<< endl;
+    totalNumberOfCupsServed = totalNumberOfCupsServed + cups;
+    cout << "total amount of tea served " << totalNumberOfCupsServed << " cups of tea"<< endl;
+    }
+    else
+        cout << "no more tea" << endl;
+    return totalNumberOfCupsServed;
+}   
+
+bool TeaParty::drink(int individuals, int cups, bool food)
+{
+    auto amountPerIndividual = cups / individuals;
+    cout << "all of the participants have had " << amountPerIndividual << " cups of tea"<< endl;
+
+    if(amountPerIndividual <= 2)
+        cout << "they can have more tea" << endl;
+    else
+    {
+        tooMuchTea = true;
+        cout << "too much tea" << endl;
+    }
+    if (food)
+        cout << "they should eat something" << endl;
+    return tooMuchTea;
+}
+
+void TeaParty::spill()
+{
+    auto amountSpilled = (totalNumberOfCupsServed / participants);
+    for(int i = 0; i < amountSpilled; i++)
+    {
+        cout << " 1 cup is spilled " << endl;
+    }
+}
 
 /*
  new UDT 4:
  */
+ struct PastaShopNewHire
+ {
+    PastaShopNewHire()
+    {
+        cout << "construct PastaShopNewHire" << endl;
+    }
+
+    ~PastaShopNewHire()
+    {
+        cout << "destruct PastaShopNewHire" << endl;
+    }
+
+    PastaShop pastaShop;
+
+    void sellNoodles(int soldAmount);
+    int packNoodles(int packages_);
+    int output();
+
+    int packages;
+    int customShapesDesigned;
+    int packagingPay;
+    int moneyMade;
+    double energy {1.0f};
+ };
+
+void PastaShopNewHire::sellNoodles(int soldAmount)
+{
+    for(int i = 0; i < soldAmount; i++)
+    {
+        pastaShop.makeCustomPasta(1.3, "New Shape", true, true, 1.0);
+    }
+}
+
+int PastaShopNewHire::packNoodles(int packages_)
+{
+    packages = packages_;
+    packagingPay = static_cast<int>(packages_ * pastaShop.pastaPrice / 3);
+    return packagingPay;
+}
+
+int PastaShopNewHire::output()
+{
+    moneyMade = pastaShop.pastaProfit + packagingPay;
+
+    for(int i = static_cast<int>(pastaShop.amountOfDoughlBS); i > 10; i--)
+    {
+       energy = energy - 0.1;
+       cout << "total energy to make pasta left " << energy << endl;
+    }
+    return moneyMade;
+}
 
 /*
  new UDT 5:
  */
+struct TrainRide
+{
+    TrainRide()
+    {
+        cout << "construct TrainRide" << endl;
+    }
+
+    ~TrainRide()
+    {
+        cout << "destruct TrainRide" << endl;
+    }
+
+    bool awake;
+    bool rain;
+    int timeAsleep; 
+    double distanceTravelled = 0.0;
+    double trainSpeedPerHour = 30.0;
+
+    double progressMade(double distance_, double trainSpeed_, int time_);
+    bool wakeUp();
+    bool goToSleep();
+
+    TeaParty teaParty{15};
+};
+
+double TrainRide::progressMade(double distance_, double trainSpeed_, int time_)
```

Comment:
underscores are reserved for the STL implementations.  Do not use underscores in the names of function arguments, member variables, etc.
- [ ] revise